### DISCUSSION TITLE
Database query in model fails silently

### DISCUSSION THREAD

**Discussion by oslabkyy** [phalcon/cphalcon#15259] [Q&A] [answered]
I am in progress of upgrading from Phalcon 3.4.2 to 4.0.1. I went through Upgrade Guideline and made changes accordingly to it. Application itself is running, I can see that all services are running and controller is called but when I'm trying to get something from the database, application fails and doesn't send any response. It fails absolutely silently, there are no errors in apache logs.

When I build the query in my model it works but it stops working right after calling `execute()` method:

```
class Users extends Model
{
    public static function getById($id) {
        return self::query()
            ->where('users_id = :users_id:')
            ->bind(
                [
                    'users_id' => $id,
                ])
            ->execute()
            ->getFirst();
    }
}
```

Decomposing the code above I can see that it crashes on execution. It also crashes on calling `Users::find()` from controller and these queries do not raise any exceptions.

I checked my DB connection by creating a simple connection in controller and realized that it works fine and returns a resultset.

Here you can see how I initialize DB connection to be used by models around the application:

```
$di->set('db', function() use ($config) {
    try {
        $dbclass = 'Phalcon\Db\Adapter\Pdo\\' . $config->database->adapter;
        return new $dbclass(array(
            'host' => $config->database->host,
            'username' => $config->database->username,
            'password' => $config->database->password,
            'dbname' => $config->database->name,
            'charset' => $config->database->charset,
        ));
    } catch (Exception $e) {
        die(sprintf("<b>Error when initializing database connection:</b> %s", $e->getMessage()));
    }
});
```

This code does not raise any exceptions too.

`var_dump(Phalcon\Di::getDefault()->get('db')); exit();` right before running query returns connection info.

I spent 2 days debugging this stuff but it's very hard when Phalcon does not provide any logs or exceptions. Does anyone have an idea what can be wrong or how can I make Phalcon speak?

**Comment by quasipickle** [comment]
Phalcon's View layer suppresses all output not through the View, so if any error messages get output, they don't get displayed.  You could try `exit()`-ing in your method to see if any problems are being hidden:

```php
public static function getById($id) {
    $result = self::query()
        ->where('users_id = :users_id:')
        ->bind(
            [
                'users_id' => $id,
            ])
        ->execute()
        ->getFirst();
    exit();
}
```
If that doesn't provide any results, you could maybe try logging directly in the database?  With that said - database issues rarely cause silent script halting.


Also, your method can be rewritten more simply with `findFirst()`
```php
public static function getById($id) {
    return self::findFirst(['users_id = :users_id:','bind'=>['users_id'=>$id]]);
}
```

**Reply by oslabkyy** [reply]
@quasipickle Of course, I used var_dump() along with `exit()` function and figured out that it crashes right after executing or even by running simply `Users::find()` method.

Also, to make sure this is not a problem with my model I created a new table with the name `testtable` with just 2 fields and a very simple model:

```<?php

use Phalcon\Mvc\Model;

class Testtable extends Model
{
    /**
     * @var integer
     */
    public $id;

    /**
     * @var string
     */
    public $name;
}
```

and by running just `Testtable::find();` from controller it crashes again silently.

My PHP version is 7.4.3, I have PDO driver and PDO extension installed. MySQL version is 8.0.22.

**Reply by quasipickle** [reply]
Hmmm.  It looks like you've been pretty diligent in trying to track down the error.  You're checking the apache `error.log` file right, not `access.log`?  You probably are - but thought I'd ask just to be safe.

Maybe try doing a raw query directly using the `db` service.  That might help narrow down whether the problem is actually with your database connection or with something in Phalcon involving models.  If that fails, try connecting using `mysqli_*` functions to see if the problem is PDO.  I'd also turn on MySQL's [general query log](https://dev.mysql.com/doc/refman/8.0/en/query-log.html) and do a `tail -f` while you try to run your queries.  That might give you some insight as to what the database server is seeing/acting.

**Reply by oslabkyy** [reply]
Can you please explain what did you mean saying to try `db` service directly?

I tried to create connection by following way and it works perfectly.
```
        $connection = new MysqlConnection(
            [
                "host"     => "localhost",
                "username" => "username",
                "password" => "password",
                "dbname"   => "dbname",
                "port"     => "3306",
            ]
        );

        $result = $connection->query(
            "SELECT * FROM testtable"
        );

        $data = $result->fetchAll();
        var_dump($data); exit();
```

Output:
`array(2) { [0]=> array(4) { ["id"]=> string(1) "1" [0]=> string(1) "1" ["name"]=> string(4) "John" [1]=> string(4) "John" } [1]=> array(4) { ["id"]=> string(1) "2" [0]=> string(1) "2" ["name"]=> string(3) "Bob" [1]=> string(3) "Bob" } }`

Thanks, I will try to check MySQL logs.

**Reply by quasipickle** [reply]
You can run a query using the `db` service like:

```php
$this->getDi()->getShared('db')->query('SELECT * FROM testtable')
```

But that's pretty much what you did.  One thing though - try making the connection while referencing the config values you used in the db service set up.  Maybe there's an errant character in there causing a problem that you won't find by manually re-typing

**Reply by oslabkyy** [reply]
Nope, `db` service is set up correctly, following code gives me results from DB:

```var_dump($this->getDi()->getShared('db')->query('SELECT * FROM testtable')->fetchAll()); exit();```

**Reply by oslabkyy** [reply]
Also, I realized that `Testtable::find();` does not log any queries in the general log, at the same time previous code does. So the problem is somewhere at Phalcon level.

**Reply by quasipickle** [reply]
Does your bootstrap file do anything with models?  Can you paste it here?

**Reply by oslabkyy** [reply]
I commented out everything which is not essential and finally, I've got this minimum configuration with the same problem.

index.php:
```php
<?php
error_reporting(E_ALL);

use Phalcon\Mvc\Application;
use Phalcon\Config\Adapter\Ini as ConfigIni;

define('APP_PATH', realpath('..') . '/');

$config = new ConfigIni(APP_PATH . 'app/config/config.ini');

require APP_PATH . 'app/config/loader.php';
require APP_PATH . 'app/config/services.php';

$application = new Application($di);
echo $application->handle($_SERVER['REQUEST_URI'])->getContent();
```

loader.php:

```php
<?php
use Phalcon\Loader;

$loader = new Loader();

$loader->registerDirs(
	array(
        APP_PATH . $config->application->controllersDir,
        APP_PATH . $config->application->modelsDir,
	)
);

$loader->register();
```

services.php
```php
<?php

use Phalcon\Mvc\View;
use Phalcon\Di\FactoryDefault;
use Phalcon\Mvc\View\Engine\Volt as VoltEngine;

$di = new FactoryDefault();

$di->set('view', function() use ($config) {

	$view = new View();

	$view->setViewsDir(APP_PATH . $config->application->viewsDir);

    $view->registerEngines(array(
        '.volt' => function ($view, $di) use ($config) {

            $volt = new VoltEngine($view, $di);
            $volt->setOptions(array(
                'compiledPath' => APP_PATH . "cache/volt/",
                'compiledSeparator' => '_',
                'compileAlways' => true
            ));

            return $volt;
        }
    ));

	return $view;
}, true);


$di->setShared('db', function() use ($config) {
    try {
        $dbclass = 'Phalcon\Db\Adapter\Pdo\\' . $config->database->adapter;
        return new $dbclass(array(
            'host' => $config->database->host,
            'username' => $config->database->username,
            'password' => $config->database->password,
            'dbname' => $config->database->name,
            'charset' => $config->database->charset
        ));
    } catch (Exception $e) {
        die(sprintf("<b>Error when initializing database connection:</b> %s", $e->getMessage()));
    }
});

$di->set('config', function() use ($config) {
    return $config;
});
```

**Reply by oslabkyy** [reply]
If this is will be helpful - my configuration:

Ubuntu 20.04.1 LTS (Focal Fossa)
PHP 7.4.3
MySQL 8.0.22
PDO Driver enabled
pdo_mysql extension installed
Phalcon 4.0.1

Phalcon config:

php -i | grep phalcon

```
/etc/php/7.4/cli/conf.d/50-phalcon.ini
phalcon
phalcon => enabled
phalcon.db.escape_identifiers => On => On
phalcon.db.force_casting => Off => Off
phalcon.orm.case_insensitive_column_map => Off => Off
phalcon.orm.cast_last_insert_id_to_int => Off => Off
phalcon.orm.cast_on_hydrate => Off => Off
phalcon.orm.column_renaming => On => On
phalcon.orm.disable_assign_setters => Off => Off
phalcon.orm.enable_implicit_joins => On => On
phalcon.orm.enable_literals => On => On
phalcon.orm.events => On => On
phalcon.orm.exception_on_failed_metadata_save => On => On
phalcon.orm.exception_on_failed_save => Off => Off
phalcon.orm.ignore_unknown_columns => Off => Off
phalcon.orm.late_state_binding => Off => Off
phalcon.orm.not_null_validations => On => On
phalcon.orm.update_snapshot_on_save => On => On
phalcon.orm.virtual_foreign_keys => On => On
phalcon.warning.enable => On => On
```

**Reply by quasipickle** [reply]
Well dang.  I don't know where else to look.  I notice you're running 4.0.1, while 4.1.0 is the latest.  Maybe try updating Phalcon itself?

**Reply by oslabkyy** [reply]
Ahh, sorry, it's just a typo, I use 4.1.0

<img width="945" alt="Screenshot 2021-01-20 at 22 31 00" src="https://user-images.githubusercontent.com/6586095/105231298-54f87a80-5b6f-11eb-80df-7900ddf76e73.png">

**Reply by quasipickle** [reply]
Double dang.

If I were in your position I would create a new project as bare bones as possible to try to replicate the problem - it should be possible to do this all in one file - then file a bug report.  If you can't replicate the problem, copy the working code in from your replication one chunk at a time until the problem goes away.

Or maybe wait until someone else chimes in with a more professional answer ;)

Good luck.

**Reply by oslabkyy** [reply]
Yeah, looks like a new clean project is the last resort in this situation that I should try.

Thank you very much for your responses and your time!

**Reply by oslabkyy** [reply]
@quasipickle btw, do you use 4.1.0 version?

**Reply by quasipickle** [reply]
No, I'm still running on 3.something.  Otherwise I'd have given your setup a try to see if I could replicate it.

**Comment by DrJLT** [comment]
Try

public static function getById($id) {
        return Di::getDefault()->getShared('db')->fetchOne('SELECT what_you_need FROM your_user_table WHERE user_id = ? LIMIT 1',\Pdo::FETCH_ASSOC,[$id]);
}

This should work.

**Reply by oslabkyy** [reply]
As I wrote above, this approach works fine, but this is a workaround. I need to use my models which is the correct way of using ORM.

**Reply by DrJLT** [reply]
why use orm? especially with a static method?

**Reply by oslabkyy** [reply]
Because it's nice, more clear, less verbose, and easily maintainable. But this topic is for another thread, I'm trying to figure out why Model does not use `'db'` service in 4.1.0 which worked perfectly in 3.4.2?

**Comment by oslabkyy** [comment]
This problem is not resolved yet. Does anyone have any ideas? Please, help!

**Reply by quasipickle** [reply]
Have you tried replicating the problem with a fresh, minimal project?

**Reply by oslabkyy** [reply]
Yes, I tried and still have the same problem. My application now is as minimal as possible.

public/index.php

```php
<?php
use Phalcon\Mvc\Application;
use Phalcon\Mvc\View;
use Phalcon\Di\FactoryDefault;
use Phalcon\Mvc\View\Engine\Volt as VoltEngine;
use Phalcon\Loader;

define('APP_PATH', realpath('..') . '/');

$loader = new Loader();

$loader->registerDirs(
    array(
        APP_PATH . 'app/controllers/',
        APP_PATH . 'app/models/',
        APP_PATH . 'app/forms/',
        APP_PATH . 'app/views/',
    )
);

$loader->register();

$di = new FactoryDefault();

$di->set('view', function() {

    $view = new View();

    $view->setViewsDir(APP_PATH . 'app/views/');

    $view->registerEngines(array(
        '.volt' => function ($view, $di) {

            $volt = new VoltEngine($view, $di);
            $volt->setOptions(array(
                'compiledPath' => APP_PATH . "cache/volt/",
                'compiledSeparator' => '_',
                'compileAlways' => true
            ));

            return $volt;
        }
    ));

    return $view;
}, true);


$di->set('db', function() {
    return new Phalcon\Db\Adapter\Pdo\Mysql(array(
        'host' => 'localhost',
        'username' => 'username',
        'password' => 'password',
        'dbname' => 'dbname',
        'charset' => 'utf8'
    ));
}, true);

$application = new Application($di);
echo $application->handle($_SERVER['REQUEST_URI'])->getContent();
```

app/controllers/IndexController.php

```php
<?php

use Phalcon\Mvc\Controller;

class IndexController extends Controller
{
    public function indexAction()
    {
        //var_dump($this->getDi()->getShared('db')->query('SELECT * FROM users')->fetchAll()); exit();
        $users = Users::find();
        var_dump($users); exit();
    }
}
```

app/models/Users.php

```php
<?php

use Phalcon\Mvc\Model;

class Users extends Model {}
```

And again, that commented `var_dump` in controller works fine but call to `find` method of model fails silently.

**Comment by Jeckerson** [comment]
Your errors might be disabled in your environment. Try to wrap with try catch inside `getById()`:

```
<?php

class Users extends Model
{
    public static function getById($id) {
        try {
            return self::query()
                ->where('users_id = :users_id:')
                ->bind(
                    [
                        'users_id' => $id,
                    ])
                ->execute()
                ->getFirst();
        } catch (\Throwable $exception) {
            echo $exception->getMessage();
            exit;
        }
    }
}
```

**Reply by oslabkyy** [reply]
Unfortunately `catch` does not catch any exceptions. The application fails silently.

**Comment by Jeckerson** [comment]
Please show output of `php -v`

**Reply by oslabkyy** [reply]
PHP 7.4.3 (cli) (built: Oct  6 2020 15:47:56) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.3, Copyright (c), by Zend Technologies

**Reply by oslabkyy** [reply]
@Jeckerson Do you have any ideas?

**Reply by Jeckerson** [reply]
@oslabkyy I noticed that you are using `4.0.1`, try to use Phalcon `v4.0.6` or `v4.1.0`.

**Reply by oslabkyy** [reply]
I use 4.1.0

**Reply by oslabkyy** [reply]
@Jeckerson magic! 4.0.6 works fine.

**Comment by oslabkyy** [comment, marked as answer]
I compiled 4.1.0 from the source code and the problem was resolved. Seems like the problem was with debian package for Ubuntu.

**Reply by Jeckerson** [reply]
Could you tell with what Zephir version you did compilation?

**Reply by oslabkyy** [reply]
0.12.19-4345593

**Reply by Jeckerson** [reply]
How did you installed Phalcon before re-compilation?

**Reply by oslabkyy** [reply]
It was installed via `apt-get install`

**Comment by zicare** [comment]
Had the same issue as oslaskyy when using Phalcon ORM with Ubuntu 20.04, PHP 7.4 and Phalcon 4.1.x. This discussion helped figure it out what the problem was. Downgrading to Phalcon 4.0.6 worked fine.

Anybody with Ubuntu 20.04 and PHP 7.4 installing Phalcon 4 following the official installation guide will likely be stuck with this problem. The workaround below can help solve it without resorting to compile the sources.

Phalcon 4.0.6 for Ubuntu 20.04 (focal) is not available at https://packagecloud.io/phalcon/stable/ubuntu/ but it is there for Ubuntu 18.04 (bionic). Manually edit file /etc/apt/sources.list.d/phalcon_stable.list

change these lines:

deb https://packagecloud.io/phalcon/stable/ubuntu/ focal main
deb-src https://packagecloud.io/phalcon/stable/ubuntu/ focal main

to:

deb https://packagecloud.io/phalcon/stable/ubuntu/ bionic main
deb-src https://packagecloud.io/phalcon/stable/ubuntu/ bionic main

Then install 4.0.6 with:

apt-get install php7.4-phalcon=4.0.6-975+php7.4

**Reply by malsksoa** [reply]
facing few issues could you please help us to [fix](https://www.deepmaterialcn.com/) it? i need your honest recommendations please?
